### PR TITLE
Feature/sg 98 add external data loaders support

### PR DESCRIPTION
**Pull request by najeeb5**
- Able to pass train/valid/test dataloaders to SgModel making the usage of DatasetInterface optional.
- Unit tests

### CODE REVIEW

File: tests/unit_tests/initialize_with_dataloaders_test.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                         "optimizer": "SGD",
+                         "optimizer_params": {"weight_decay": 1e-5, "momentum": 0.9},
+                         "train_metrics_list": [Accuracy()],
+                         "valid_metrics_list": [Accuracy()],
+                         "metric_to_watch": "Accuracy",
+                         "greater_metric_to_watch_is_better": True})
+        self.assertTrue(0 < model.best_metric.item() < 1)
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (shaydeci) [inline tests/unit_tests/initialize_with_dataloaders_test.py:84]: make sure you run linter locally before pushing as we still need to add it to ci.

**Review by shaydeci — CHANGES_REQUESTED** [review]
Some small changes stated inline..
Nice work! Especially on the unit tests!

**Comment by shaydeci** [comment]
Additionally, I think it'd be good idea to demonstrate this feature as a part of how one can train with their own initialised objects.
My suggestion- let's rename the cifar10_external_optimizer_example.py which almost entirely uses initialised objects to cifar10_initilized_objects_example.py and show how one can also pass data loaders there.
you can also be a good summerian and pass an initialised loss function so it'd completely run with external objects.

**Review by shaydeci — APPROVED** [review]
LGTM

File: src/super_gradients/training/sg_model/sg_model.py [hunk 2]
```diff
@@ -221,11 +221,19 @@ def __init__(self, experiment_name: str, device: str = None, multi_gpu: Union[Mu
...
         dataset_params = {"batch_size": train_loader.batch_size if train_loader else None,
                           "val_batch_size": valid_loader.batch_size if valid_loader else None,
                           "test_batch_size": test_loader.batch_size if test_loader else None,
                           "dataset_dir": None,
                           "s3_link": None}
+
+        if train_loader and self.multi_gpu == MultiGPUMode.DISTRIBUTED_DATA_PARALLEL:
+            if not all([isinstance(train_loader.sampler, DistributedSampler),
+                        isinstance(valid_loader.sampler, DistributedSampler),
+                        isinstance(test_loader.sampler, DistributedSampler) or test_loader is None]):
```
Comment (shaydeci) [inline src/super_gradients/training/sg_model/sg_model.py:234]: Dod you check this warns..? Im pretty sure isinstance(test_loader.sampler, DistributedSampler) or test_loader is None will through an AttributeError if test sampler is None.

**Review by shaydeci — CHANGES_REQUESTED** [review]
short comment inline

**Review by shaydeci — APPROVED** [review]
re LGTM

**Review by shaydeci — APPROVED** [review]
